lib/rules/no-shadow.js

```diff
@@ -75,10 +109,176 @@ module.exports = {
 	},
 
 	create(context) {
-		const [{ builtinGlobals, hoist, allow, ignoreOnInitialization }] =
-			context.options;
+		const [
+			{
+				builtinGlobals,
+				hoist,
+				allow,
+				ignoreOnInitialization,
+				ignoreTypeValueShadow,
+				ignoreFunctionTypeParameterNameValueShadow,
+			},
+		] = context.options;
 		const sourceCode = context.sourceCode;
 
+		/**
+		 * Check if a scope is a TypeScript module augmenting the global namespace.
+		 * @param {Scope} scope The scope to check
+		 * @returns {boolean} Whether the scope is a global augmentation
+		 */
+		function isGlobalAugmentation(scope) {
+			return (
+				scope.block.kind === "global" ||
+				(!!scope.upper && isGlobalAugmentation(scope.upper))
+			);
+		}
+
+		/**
+		 * Check if variable is a `this` parameter.
+		 * @param {Object} variable The variable to check
+		 * @returns {boolean} Whether the variable is a this parameter
+		 */
+		function isThisParam(variable) {
+			return (
+				variable.defs[0].type === "Parameter" &&
+				variable.name === "this"
+			);
+		}
+
+		/**
+		 * Checks if type and value shadows each other
+		 * @param {Object} variable The variable to check
+		 * @param {Object} shadowedVariable The shadowed variable
+		 * @returns {boolean} Whether it's a type/value shadow case to ignore
+		 */
+		function isTypeValueShadow(variable, shadowedVariable) {
+			if (ignoreTypeValueShadow !== true) {
+				return false;
+			}
+
+			if (!("isValueVariable" in variable)) {
+				return false;
+			}
+
+			const firstDefinition = shadowedVariable.defs[0];
+
+			// Check if shadowedVariable is a type import
+			const isTypeImport =
+				firstDefinition &&
+				firstDefinition.parent?.type === "ImportDeclaration" &&
+				(firstDefinition.parent.importKind === "type" ||
+					firstDefinition.parent.specifiers.some(
+						s => s.importKind === "type",
+					));
+
+			const isShadowedValue =
+				!firstDefinition ||
+				(isTypeImport ? false : shadowedVariable.isValueVariable);
+
+			return variable.isValueVariable !== isShadowedValue;
+		}
+
+		/**
+		 * Checks if it's a function type parameter shadow
+		 * @param {Object} variable The variable to check
+		 * @returns {boolean} Whether it's a function type parameter shadow case to ignore
+		 */
+		function isFunctionTypeParameterNameValueShadow(variable) {
+			if (ignoreFunctionTypeParameterNameValueShadow !== true) {
+				return false;
+			}
+
+			return variable.defs.every(def =>
+				ALLOWED_FUNCTION_VARIABLE_DEF_TYPES.has(def.node.type),
+			);
+		}
+
+		/**
+		 * Checks if the variable is a generic of a static method
+		 * @param {Object} variable The variable to check
+		 * @returns {boolean} Whether the variable is a generic of a static method
+		 */
+		function isGenericOfStaticMethod(variable) {
```

Comment:
[Naming] Since this code was originally written, the general consensus on "generics" has shifted: https://www.totaltypescript.com/no-such-thing-as-a-generic. I think this would be a more precise name in line with what we'd normally call a type parameter:

```suggestion
		function isTypeParameterOfStaticMethod(variable) {
```